core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1523,22 +1523,21 @@ abstract class RDD[T: ClassTag](
    * @return an array of top elements
    */
   def takeOrdered(num: Int)(implicit ord: Ordering[T]): Array[T] = withScope {
-    if (num == 0) {
+    if (num == 0 || this.getNumPartitions == 0) {
       Array.empty
     } else {
-      val mapRDDs = mapPartitions { items =>
-        // Priority keeps the largest elements, so let's reverse the ordering.
-        val queue = new BoundedPriorityQueue[T](num)(ord.reverse)
-        queue ++= collectionUtils.takeOrdered(items, num)(ord)
-        Iterator.single(queue)
-      }
-      if (mapRDDs.partitions.length == 0) {
-        Array.empty
-      } else {
-        mapRDDs.reduce { (queue1, queue2) =>
-          queue1 ++= queue2
-          queue1
-        }.toArray.sorted(ord)
+      this.mapPartitionsWithIndex { case (pid, iter) =>
+        if (iter.nonEmpty) {
+          // Priority keeps the largest elements, so let's reverse the ordering.
+          Iterator.single(collectionUtils.takeOrdered(iter, num)(ord).toArray)
+        } else if (pid == 0) {
+          // make sure partition 0 always returns an array to avoid reduce on empty RDD
+          Iterator.single(Array.empty[T])
+        } else {
+          Iterator.empty
+        }
+      }.reduce { (array1, array2) =>
+        collectionUtils.mergeOrdered[T](Seq(array1, array2))(ord).take(num).toArray
```

Comment:
Nit: you can statically create the array upfront (`min(num, array1.length + array2.length)` ) and copy into it, instead of `toArray`

Reply:
good point

Reply:
Wondering if there was any impact on the perf numbers :-)

Reply:
When both `num` and  `numParitions` are large, I think we should use `treeReduce` instead, then there will be no bottleneck at driver.
I'm not sure whether it is worthwhile to add anther configuration for it.

Reply:
if the concern is that one new array will be created at each merge, I guess we can switch to an [in-place merge sort](https://www.techiedelight.com/inplace-merge-two-sorted-arrays/)